File: backend/app/schemas/organism.py
from __future__ import annotations

import uuid

from app.models.enums import OrganismGroup
from app.schemas.common import CamelModel


class OrganismRead(CamelModel):
    id: uuid.UUID
    slug: str
    scientific_name: str
    common_name: str | None = None
    tax_id: int
    rank: str | None = None
    lineage: list[str] = []
    group: OrganismGroup
    image_url: str | None = None
    sequence_count: int | None = None
    extinction_status: str | None = None
    extinction_date_text: str | None = None
    geologic_period: str | None = None
    paleogenomic_slug: str | None = None


class OrganismListResponse(CamelModel):
    organisms: list[OrganismRead]
    total: int
    next_cursor: str | None = None

File: frontend/src/types/organism.ts
import type { CategoryKey } from "@/lib/design-tokens";

export type OrganismGroup =
  | "animal"
  | "plant"
  | "fungus"
  | "bacteria"
  | "archaea"
  | "virus"
  | "protozoan";

export interface OrganismLink {
  label: string;
  url: string;
  external?: boolean;
}

export interface Organism {
  id: string;
  slug: string;
  scientificName: string;
  commonName?: string;
  /** NCBI Taxonomy ID — the canonical, verifiable identifier. */
  taxId: number;
  rank: string;
  /** Ordered lineage from broad to specific. */
  lineage: string[];
  group: OrganismGroup;
  category?: CategoryKey;
  /**
   * Real record count from the backend. `null` means "not yet available" — the
   * UI shows a neutral state instead of ever inventing a number.
   */
  sequenceCount: number | null;
  /** Provided by the backend/media service; null renders the abstract emblem. */
  imageUrl?: string | null;
  links: OrganismLink[];
  extinctionStatus?: string | null;
  extinctionDateText?: string | null;
  geologicPeriod?: string | null;
  paleogenomicSlug?: string | null;
}
